c7n/ctx.py

```diff
@@ -26,43 +35,87 @@ def __init__(self, session_factory, policy, options):
         self.policy = policy
         self.options = options
         self.session_factory = session_factory
-        self.cloudwatch_logs = None
+
+        # Runtime initialized during policy execution
+        # We treat policies as a fly weight pre-execution.
         self.start_time = None
+        self.execution_id = None
+        self.output = None
+        self.api_stats = None
+        self.sys_stats = None
+
+        # A few tests patch on metrics flush
+        self.metrics = metrics_outputs.select(self.options.metrics_enabled, self)
```

Comment:
one alternatives come to mind, an initialized boolean field on execution context that allows a context to only be initialized once, that tests that need this can preinitialize a policy before run so they can patch before run.